### ISSUE TITLE
Speaker Diarization

### ISSUE THREAD

**Issue by kaustavb79** [NVIDIA/NeMo#4157]
I am trying to implement speaker diarization using __vad_marblenet__ and __titanet_large__ models.

# Issue
When I am trying to run on server/local machine .... the process either gets stuck or terminates due to CPU memory allocation error during clustering.
- __memory allocation issue__
![Screenshot from 2022-05-11 17-34-38](https://user-images.githubusercontent.com/29993931/168004643-afd81bdc-d018-422a-9a7c-1788f733d4e0.png)
- __execution halted__
![Screenshot from 2022-05-12 12-01-30](https://user-images.githubusercontent.com/29993931/168006552-788685e5-be35-4ee2-8035-9b4e11517c51.png)
![Screenshot from 2022-05-12 12-03-23](https://user-images.githubusercontent.com/29993931/168006635-73c1b165-4e95-43f8-90c2-c38fd1edcd64.png)

## Environment:
- OS : Ubuntu 18.04.6 LTS (GNU/Linux 4.18.0-25-generic x86_64)
- Memory: 
![Screenshot from 2022-05-12 11-52-18](https://user-images.githubusercontent.com/29993931/168005126-f3197681-38c3-4b6a-834d-d5fe5df053a5.png)
- Python: 3.7.13
- Installation : from source i.e. __(git clone ... & cd NeMo & ./reinstall.sh)__
# Reference
When executing on collab, it's working properly. It's giving proper diarization results.
 - Link to notebook: https://colab.research.google.com/drive/1UgzoPHV1aByDDhygCUAiEaEIRGldrocK?usp=sharing
 - Test audio file: https://drive.google.com/file/d/1ChLsVHTgH4SUgvqLP7ymtrbao-L0KKsj/view?usp=sharing
 - Manifest file: https://drive.google.com/file/d/1SAeLmMIHGk-rc7ydHfYLywCkbMx7SbH6/view?usp=sharing
 - Diarization yaml config file: https://drive.google.com/file/d/1lvYYg5Pd8hUceS2a-raO76huu-p0p39V/view?usp=sharing
 - Test audio file result: https://drive.google.com/drive/folders/1FCH_UOFUFJZUQdxMHQfzXQII-bg6auaO?usp=sharing

## Environment:
- Installation on collab: from source i.e. __(git clone ... & cd NeMo & ./reinstall.sh)__
- Python: 3.7.13

**Comment by nithinraok** [comment]
Are you running only on CPU?

**Comment by nithinraok** [comment]
I ran the particular file you mentioned: https://drive.google.com/file/d/1ChLsVHTgH4SUgvqLP7ymtrbao-L0KKsj/view?usp=sharing. I don't get any errors as you mentioned both on CPU and GPU. 

Here are two things you might missed:
* convert audio from flac to wav
* resample file from 44.1kHz to 16kHz 

Can you retry after considering above suggestions?

**Comment by kaustavb79** [comment]
> 

yes

**Comment by kaustavb79** [comment]
I am resampling the flack file to 16kHz in the code itself .

**Comment by lyceumman** [comment]
same issue during CPU inference on Ubuntu 20.04.4 LTS on Python 3.8.12 and nemo-toolkit==1.9.0rc0. Previously resampled to 16kHz my wav files. And then stuck at this moment.

<img width="1367" alt="Screenshot 2022-05-18 at 18 11 40" src="https://user-images.githubusercontent.com/37379663/169075982-910b218e-9574-4b63-9719-966780c6ace7.png">

**Comment by nithinraok** [comment]
Can you create a python script or notebook just with the file you mentioned to reproduce the error and share it here? I don't see any such issue with `capital-trade01.wav`

**Comment by kaustavb79** [comment]
[diarization_code.zip](https://github.com/NVIDIA/NeMo/files/8718092/diarization_code.zip)

**Comment by nithinraok** [comment]
We dont test using virtual env.
Please use conda and try again with min python version as 3.8

I executed your script and it ran flawlessly.

**Comment by nithinraok** [comment]
I have created a sample notebook for you on colab, upload the `capital-trade01.flac` to the notebook and it does the rest.
https://colab.research.google.com/drive/1w_tB-1hHiMCDpnlMM4ygIxNM_GM0G4Ep?usp=sharing

**Comment by kaustavb79** [comment]
> I have created a sample notebook for you on colab, upload the `capital-trade01.flac` to the notebook and it does the rest. https://colab.research.google.com/drive/1w_tB-1hHiMCDpnlMM4ygIxNM_GM0G4Ep?usp=sharing

okay

**Comment by togousch** [comment]
> Can you create a python script or notebook just with the file you mentioned to reproduce the error and share it here? I don't see any such issue with `capital-trade01.wav`

https://drive.google.com/drive/folders/1xj1GwaluJZelnG5-428rcLbsOByA-WM7?usp=sharing

Here is folder with audio files (38 .wav), i try to inference on (CPU as a device). There is script with config i use (just pretrained vad_marblenet.nemo as vad model and titanet-l.nemo as speaker model). I launch it from PyTorch container (22.03) with:
nemo-toolkit==1.9.0rc0
Python 3.8.12
Ubuntu 20.04.4 LTS

<img width="1357" alt="Screenshot 2022-05-22 at 16 37 22" src="https://user-images.githubusercontent.com/37379663/169697983-cee1ab64-acbf-44ca-b46b-9caafd9ab567.png">

By launching it i get stuck after VAD, as i mentioned earlier or even exit without throwing any error, that i can try/except on.

**Comment by nithinraok** [comment]
Could you mention the installation process you followed? 
Also these are the version I suggest: 
22.03 PyTorch docker container: https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch/tags 
NeMo branch: https://github.com/NVIDIA/NeMo/tree/v1.8.2

**Comment by togousch** [comment]
> Could you mention the installation process you followed? Also these are the version I suggest: 22.03 PyTorch docker container: https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch/tags NeMo branch: https://github.com/NVIDIA/NeMo/tree/v1.8.2

Build an image from repo https://github.com/NVIDIA/NeMo/tree/v1.8.2 with
docker build -f Dockerfile -t nemo:latest .

**Comment by EmreOzkose** [comment]
It is not about diarization, it is about voice activity detection. I also encountered this issue. Steps:
- Clone latest [NeMo repo](https://github.com/NVIDIA/NeMo/releases/tag/v1.9.0)
- cd NeMo/examples/asr/speech_classification
- create manifest.json ({"audio_filepath": "example.wav", "offset": 0, "label": "infer", "text": "-", "num_speakers": null, "uem_filepath": null})
- run `python vad_infer.py`
- It is freezing during `generate_overlap_vad_seq`

I tried trimmed audio files. I trim audio with `sox example.wav example_30s.wav trim 0 30`. Script is running without any problem if duration is less than 30, but if duration increase to 35 seconds, it is freezing

**Comment by fayejf** [comment]
This is probably due to unsolved python deadlock on multiprocessing in some environments as described in https://github.com/NVIDIA/NeMo/issues/4315. Thanks @gkucsko for providing a fix to disable it in  https://github.com/NVIDIA/NeMo/pull/4317

**Comment by togousch** [comment]
For now I've solved this problem by deploying this container on the local machine, because I used to run it on a virtual machine (docker container inside vm). This way it doesn't stuck even with thousands of long audios for me.

**Comment by tango4j** [comment]
@togousch Could you clarify two things?
(1) Are you running vad_infer.py on jupyter notebook? or running it on bash? 
(2) Could you also check `num_worker` in yaml file setting?  and please check whether num_worker=1 sovles the hanging problem?
